### PR TITLE
CoreTiming: Keep Cores synced

### PR DESCRIPTION
**Pull request by B3n30**
Keep Cores synced by having each core it's own downcount and let the main core run an intermediate run if it's slice_length was cropped

This is untested again, and has probably a lot of bugs. So please test and give me feadback

### CODE REVIEW

File: src/core/core_timing.h [hunk 1]
```diff
@@ -82,6 +81,8 @@ void MoveEvents();
 /// Pretend that the main CPU has executed enough cycles to reach the next event.
 void Idle();
 
+bool MainSliceWasCropped();
```
Comment (lioncash) [inline src/core/core_timing.h]: This should have a documentation comment just like the functions above and below it.

File: src/core/core_timing.cpp [hunk 2]
```diff
@@ -193,11 +200,19 @@ void MoveEvents() {
 }
 
 void Advance() {
+    std::unique_lock<std::mutex> lock(mutex);
+    size_t current_core = Core::System::GetInstance().CurrentCoreIndex();
```
Comment (lioncash) [inline src/core/core_timing.cpp]: `const size_t`

File: src/core/core_timing.cpp [hunk 3]
```diff
@@ -176,11 +179,15 @@ void RemoveNormalAndThreadsafeEvent(const EventType* event_type) {
...
 void ForceExceptionCheck(s64 cycles) {
     cycles = std::max<s64>(0, cycles);
-    if (downcount > cycles) {
+    std::unique_lock<std::mutex> lock(mutex);
+    if (downcount[0] > cycles) {
         // downcount is always (much) smaller than MAX_INT so we can safely cast cycles to an int
         // here. Account for cycles already executed by adjusting the g.slice_length
-        slice_length -= downcount - static_cast<int>(cycles);
-        downcount = static_cast<int>(cycles);
+        s64 crop_time = downcount[0] - static_cast<int>(cycles);
```
Comment (lioncash) [inline src/core/core_timing.cpp]: `const s64`

File: src/core/core_timing.cpp [hunk 4]
```diff
@@ -89,10 +92,11 @@ void UnregisterAllEvents() {
 }
 
 void Init() {
-    downcount = MAX_SLICE_LENGTH;
+    std::unique_lock<std::mutex> lock(mutex);
+    std::fill(downcount.begin(), downcount.end(), MAX_SLICE_LENGTH);
```
Comment (lioncash) [inline src/core/core_timing.cpp]:
```cpp
downcount.fill(MAX_SLICE_LENGTH);
```

**Comment by B3n30** [comment]
All events scheduled from another core then the main core will always have a delay (with max being MAX_SLICE_LENGTH) and I can't see any way to fix this. 

In my opinion this is the best approach to sync multiple cores without destroying performance completely

**Comment by FernandoS27** [comment]
Seems to be crashing as of  4d7e166 master.

**Comment by B3n30** [comment]
Sorry forgot about that PR.
CoreTiming needs a major rewrite to work properly with the multicore implementation of yuzu.

**Pull request closed without merging**